packages/oidc/src/token.ts

```diff
@@ -0,0 +1,226 @@
+    case 'linux':
+      return path.join('~/.local/share', vercelFolder);
+    case 'win32':
+      return path.join('%LOCALAPPDATA%', vercelFolder);
+    default:
+      return null;
+  }
+}
+
+async function getVercelCliToken(): Promise<string | null> {
+  const dataDir = await getVercelDataDir();
+  if (!dataDir) {
+    return null;
+  }
+  const path = await import('path');
+  const fs = await import('fs');
+  const tokenPath = path.join(dataDir, 'auth.json');
+  if (!fs.existsSync(tokenPath)) {
+    return null;
+  }
+  const token = fs.readFileSync(tokenPath, 'utf8');
+  return JSON.parse(token).token;
+}
+
+interface VercelTokenResponse {
+  token: string;
+}
+
+async function getVercelOidcToken(
+  authToken: string,
+  projectId: string
+): Promise<VercelTokenResponse | null> {
+  try {
+    const url = `https://api.vercel.com/projects/token?projectId=${projectId}&source=vercel-oidc-refresh`;
+    const res = await fetch(url, {
+      method: 'POST',
+      headers: {
+        Authorization: `Bearer ${authToken}`,
+      },
+    });
+    if (!res.ok) {
+      throw new Error(`failed to refresh token: ${res.statusText}`);
+    }
+    if (!res.json) {
+      throw new Error('failed to refresh token: no json response');
+    }
+    const tokenRes = (await res.json()) as VercelTokenResponse;
+    return tokenRes;
+  } catch (e) {
+    const message = e instanceof Error ? e.message : e;
+    throw new Error(`failed to refresh token: ${message}`);
+  }
+}
+
+async function findProjectId(): Promise<string> {
+  const dir = await findRootDir();
+  if (!dir) {
+    throw new Error('unable to find root directory');
+  }
+  try {
+    const fs = await import('fs');
+    const path = await import('path');
+    const pkgPath = path.join(dir, '.vercel', 'project.json');
+    if (!fs.existsSync(pkgPath)) {
+      throw new Error('project.json not found');
+    }
+    const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf8'));
+    return pkg.projectId;
+  } catch (e) {
+    const message = e instanceof Error ? e.message : e;
+    throw new Error(`unable to find project id: ${message}`);
+  }
+}
+
+async function findRootDir(): Promise<string> {
+  try {
+    const fs = await import('fs');
+    const path = await import('path');
+    let dir = process.cwd();
+    while (dir !== path.dirname(dir)) {
+      const pkgPath = path.join(dir, '.vercel');
+      if (fs.existsSync(pkgPath)) {
+        return dir;
+      }
+      dir = path.dirname(dir);
+    }
+  } catch (e) {
+    throw new Error('token refresh only supported in node server environments');
+  }
+  throw new Error('unable to find root directory');
+}
+
+async function saveToken(
+  token: VercelTokenResponse,
+  projectId: string
+): Promise<void> {
+  try {
+    const os = await import('os');
+    const path = await import('path');
+    const tmpDir = os.tmpdir();
+    const tokenPath = path.join(
+      tmpDir,
+      `vercel-oidc-token-${projectId}`,
+      'token.json'
```

Comment:
I wonder if just `${projectId}.json` would be sufficient? The project ID is prefixed `prj_`, right? So it's pretty unambiguous. WDYT?

Also, I started thinking some more about the `/tmp` directory… maybe it would be better to not choose new directories to write this information to, and to instead re-use the existing `~/Library/Application Support`, `$XDG_DATA_HOME`, etc., directory, but with `com.vercel.oidc` or something like that? In that case, `~/Library/Application Support/com.vercel.oidc/${projecId}.json` seems pretty clear? WDYT?

Reply:
I'm mostly impartial honestly. There's an argument to be made for letting those tokens get wiped when you reboot since they're for development but also the standard is already to store it in XDG_DATA_HOME so I'm up for changing it